File: internal/ast/formatter.go
package ast

import (
	"fmt"
	"strings"
)

func FormatAST(file File) string {
	var out strings.Builder
	out.WriteString("File\n")

	for _, declaration := range file.Declarations {
		formatDeclaration(&out, declaration, 1)
	}

	return out.String()
}

func formatDeclaration(
	out *strings.Builder,
	declaration Declaration,
	depth int,
) {
	switch declaration := declaration.(type) {
	case Comment:
		formatComment(out, declaration, depth)
	case *Comment:
		formatComment(out, *declaration, depth)
	case FunctionDeclaration:
		formatFunctionDeclaration(out, declaration, depth)
	case *FunctionDeclaration:
		formatFunctionDeclaration(out, *declaration, depth)
	case ConstDeclaration:
		formatConstDeclaration(out, declaration, depth)
	case *ConstDeclaration:
		formatConstDeclaration(out, *declaration, depth)
	default:
		writeLine(out, depth, "UnknownDeclaration")
	}
}

func formatComment(out *strings.Builder, comment Comment, depth int) {
	kind := "LineComment"
	if comment.Multiline {
		kind = "BlockComment"
	}

	writeLine(out, depth, "%s text=%q", kind, comment.Text)
}

func formatConstDeclaration(
	out *strings.Builder,
	declaration ConstDeclaration,
	depth int,
) {
	writeLine(out, depth, "ConstDeclaration")
	writeLine(out, depth+1, "Name")
	formatExpression(out, declaration.Name, depth+2)
	if declaration.Type.Name.Name != "" {
		writeLine(out, depth+1, "Type")
		formatExpression(out, declaration.Type.Name, depth+2)
	}
	writeLine(out, depth+1, "Value")
	formatExpression(out, declaration.Value, depth+2)
}

func formatFunctionDeclaration(
	out *strings.Builder,
	declaration FunctionDeclaration,
	depth int,
) {
	writeLine(
		out,
		depth,
		"FunctionDeclaration name=%q",
		declaration.Name,
	)
	writeLine(out, depth+1, "Parameters")
	for index, parameter := range declaration.Parameters {
		writeLine(out, depth+2, "Parameter %d", index)
		writeLine(out, depth+3, "Name")
		formatExpression(out, parameter.Name, depth+4)
		writeLine(out, depth+3, "Type")
		formatExpression(out, parameter.Type.Name, depth+4)
	}

	writeLine(out, depth+1, "ReturnTypes")
	formatTypeReferences(out, declaration.ReturnTypes, depth+2)

	if len(declaration.ErrorTypes) > 0 {
		writeLine(out, depth+1, "ErrorTypes")
		formatTypeReferences(out, declaration.ErrorTypes, depth+2)
	}

	if declaration.Body == nil {
		writeLine(out, depth+1, "Body <nil>")
		return
	}

	formatBlockStatement(out, declaration.Body, depth+1)
}

func formatTypeReferences(
	out *strings.Builder,
	types []TypeReference,
	depth int,
) {
	for index, typ := range types {
		writeLine(out, depth, "Type %d", index)
		formatExpression(out, typ.Name, depth+1)
	}
}

func formatBlockStatement(
	out *strings.Builder,
	block *BlockStatement,
	depth int,
) {
	writeLine(out, depth, "BlockStatement")

	for _, statement := range block.Statements {
		formatStatement(out, statement, depth+1)
	}
}

func formatStatement(out *strings.Builder, statement Statement, depth int) {
	switch statement := statement.(type) {
	case Comment:
		formatComment(out, statement, depth)
	case *Comment:
		formatComment(out, *statement, depth)
	case ReturnStatement:
		writeLine(out, depth, "ReturnStatement")
		for index, value := range statement.Values {
			writeLine(out, depth+1, "Value %d", index)
			formatExpression(out, value, depth+2)
		}
	case VariableDeclarationStatement:
		kind := "const"
		if statement.Mutable {
			kind = "let"
		}

		writeLine(
			out,
			depth,
			"VariableDeclarationStatement kind=%q name=%q",
			kind,
			statement.Name,
		)
		if statement.Type.Name.Name != "" {
			writeLine(out, depth+1, "Type")
			formatExpression(out, statement.Type.Name, depth+2)
		}
		writeLine(out, depth+1, "Value")
		formatExpression(out, statement.Value, depth+2)
	case ExpressionStatement:
		writeLine(out, depth, "ExpressionStatement")
		formatExpression(out, statement.Value, depth+1)
	case AssignmentStatement:
		writeLine(out, depth, "AssignmentStatement")
		writeLine(out, depth+1, "Target")
		formatExpression(out, statement.Target, depth+2)
		writeLine(out, depth+1, "Value")
		formatExpression(out, statement.Value, depth+2)
	case IfStatement:
		writeLine(out, depth, "IfStatement")
		writeLine(out, depth+1, "Condition")
		formatExpression(out, statement.Condition, depth+2)
		writeLine(out, depth+1, "Then")
		formatBlockStatement(out, &statement.ThenBlock, depth+2)
		if len(statement.ElseBlock.Statements) > 0 {
			writeLine(out, depth+1, "Else")
			formatBlockStatement(out, &statement.ElseBlock, depth+2)
		}
	case WhileStatement:
		writeLine(out, depth, "WhileStatement")
		writeLine(out, depth+1, "Condition")
		formatExpression(out, statement.Condition, depth+2)
		writeLine(out, depth+1, "Body")
		formatBlockStatement(out, &statement.Body, depth+2)
	default:
		writeLine(out, depth, "UnknownStatement")
	}
}

func formatExpression(out *strings.Builder, expression Expression, depth int) {
	switch expression := expression.(type) {
	case UnaryExpression:
		writeLine(out, depth, "UnaryExpression operator=%q", expression.Operator)
		writeLine(out, depth+1, "Expression")
		formatExpression(out, expression.Expression, depth+2)
	case BinaryExpression:
		writeLine(out, depth, "BinaryExpression operator=%q", expression.Operator)
		writeLine(out, depth+1, "Left")
		formatExpression(out, expression.Left, depth+2)
		writeLine(out, depth+1, "Right")
		formatExpression(out, expression.Right, depth+2)
	case FunctionCallExpression:
		writeLine(out, depth, "FunctionCallExpression")
		writeLine(out, depth+1, "Callee")
		formatExpression(out, expression.Callee, depth+2)
		writeLine(out, depth+1, "Arguments")
		for index, argument := range expression.Arguments {
			writeLine(out, depth+2, "Argument %d", index)
			formatExpression(out, argument, depth+3)
		}
	case IntegerLiteral:
		writeLine(out, depth, "IntegerLiteral value=%q", expression.Value)
	case BooleanLiteral:
		writeLine(out, depth, "BooleanLiteral value=%q", expression.Value)
	case StringLiteral:
		writeLine(out, depth, "StringLiteral value=%q", expression.Value)
	case CharacterLiteral:
		writeLine(out, depth, "CharacterLiteral value=%q", expression.Value)
	case Identifier:
		writeLine(out, depth, "Identifier name=%q", expression.Name)
	default:
		writeLine(out, depth, "UnknownExpression")
	}
}

func writeLine(out *strings.Builder, depth int, format string, args ...any) {
	out.WriteString(strings.Repeat("  ", depth))
	fmt.Fprintf(out, format, args...)
	out.WriteByte('\n')
}

File: src/ast/formatter.ts
import { convert } from "@catalystic/json-to-yaml";
import type {
    AssignmentStatement,
    BlockStatement,
    CheckBlockStatement,
    Declaration,
    EnumDecl,
    Expression,
    ExpressionStatement,
    ForStatement,
    ForwardStatement,
    FunctionDeclaration,
    FunctionParameter,
    Identifier,
    IfStatement,
    ImportDeclaration,
    InterfaceDeclaration,
    InterfaceMethodRequirement,
    Module,
    ObjectLiteralElement,
    ObjectLiteralExpression,
    ReturnStatement,
    ReturnErrorStatement,
    Statement,
    StructDecl,
    SwitchCase,
    SwitchStatement,
    Type,
    TypeDeclaration,
    UnionDecl,
} from "./types.js";
import { TypeDeclKind } from "./types.js";

/**
 * Renders a parsed {@link Module} into a simplified, human-readable tree.
 *
 * The formatter walks the AST and rebuilds each node as a plain object that
 * drops compiler-internal detail (notably source {@link Position}s), producing
 * a stable shape suitable for printing as YAML when inspecting or snapshotting
 * the parser's output.
 */
export class Formatter {
    ast: Module;

    constructor(ast: Module) {
        this.ast = ast;
    }

    formatAssignmentStatement(e: AssignmentStatement) {
        return {
            kind: "assignment_statement",
            root: this.formatExpression(e.root),
            target: this.formatExpression(e.target),
            ...(e.asResult ? { asResult: e.asResult.resultName.name } : {}),
        };
    }

    formatForStatement(e: ForStatement): any {
        return {
            kind: "for_statement",
            ...(e.declaration ? { declaration: this.formatStatement(e.declaration) } : {}),
            ...(e.condition ? { condition: this.formatExpression(e.condition) } : {}),
            ...(e.modifier ? { modifier: this.formatExpression(e.modifier) } : {}),
            body: this.formatBlockStatement(e.body),
        };
    }

    formatIfStatement(e: IfStatement): any {
        return {
            kind: "if_statement",
            condition: this.formatExpression(e.condition),
            thenBlock: this.formatBlockStatement(e.thenBlock),
            ...(e.elseBlock ? { elseBlock: this.formatBlockStatement(e.elseBlock) } : {}),
        };
    }

    /** Formats a single `case`/`default` arm: its labels and body statements. */
    formatSwitchCase(c: SwitchCase): any {
        return {
            labels: c.labels.map((x) => this.formatExpression(x)),
            body: c.body.statements.map((x) => this.formatStatement(x)),
        };
    }

    formatSwitchStatement(e: SwitchStatement): any {
        return {
            kind: "switch_statement",
            scrutinee: this.formatExpression(e.scrutinee),
            cases: e.cases.map((x) => this.formatSwitchCase(x)),
            ...(e.default ? { default: this.formatSwitchCase(e.default) } : {}),
        };
    }

    /** Formats an object-literal field or spread while dropping source positions. */
    formatObjectLiteralElement(element: ObjectLiteralElement): any {
        switch (element.kind) {
            case "field_init":
                return {
                    kind: "field_init",
                    field: {
                        name: this.formatIdentifier(element.field.name),
                        value: this.formatExpression(element.field.value),
                    },
                };
            case "spread_element":
                return {
                    kind: "spread_element",
                    source: this.formatExpression(element.source),
                };
        }
    }

    /** Formats a record/object literal and recursively formats its elements. */
    formatObjectLiteralExpression(expression: ObjectLiteralExpression): any {
        return {
            kind: "object_literal",
            type: this.formatType(expression.type),
            ...(expression.genericTypes?.length
                ? { typeArguments: expression.genericTypes.map((type) => this.formatType(type)) }
                : {}),
            elements: expression.elements.map((element) =>
                this.formatObjectLiteralElement(element),
            ),
        };
    }

    /** Formats a single expression node. */
    formatExpression(e?: Expression): any {
        if (!e) {
            return {};
        }
        switch (e.kind) {
            case "new_expression":
                return {
                    kind: "new_expression",
                    expression: this.formatExpression(e.expression),
                };
            case "object_literal":
                return this.formatObjectLiteralExpression(e);
            case "identifier":
                return this.formatIdentifier(e);
            case "integer_literal":
                return {
                    kind: "integer_literal",
                    value: e.value,
                };
            case "float_literal":
                return {
                    kind: "float_literal",
                    value: e.value,
                };
            case "boolean_literal":
                return {
                    kind: "boolean_literal",
                    value: e.value,
                };
            case "char_literal":
                return {
                    kind: "char_literal",
                    value: e.value,
                };
            case "string_literal":
                return {
                    kind: "string_literal",
                    value: e.value,
                };
            case "array_literal_expression":
                return {
                    kind: "array_literal_expression",
                    elements: e.elements.map((x) => this.formatExpression(x)),
                };
            case "index_expression":
                return {
                    kind: "index_expression",
                    receiver: this.formatExpression(e.receiver),
                    index: this.formatExpression(e.index),
                };
            case "function_call_expression":
                return {
                    kind: "function_call_expression",
                    conversion: e.conversion,
                    callee: this.formatExpression(e.callee),
                    ...(e.genericTypes?.length
                        ? { typeArguments: e.genericTypes.map((x) => this.formatType(x)) }
                        : {}),
                    arguments: e.arguments.map((x) => this.formatExpression(x)),
                };
            case "unary_expression":
                return {
                    kind: "unary_expression",
                    operator: e.operator,
                    operand: this.formatExpression(e.operand),
                };
            case "binary_expression":
                return {
                    kind: "binary_expression",
                    left: this.formatExpression(e.left),
                    operator: e.operator,
                    right: this.formatExpression(e.right),
                };
        }
    }

    /** Formats a single statement, dispatching on its `kind`. */
    formatStatement(b: Statement): any {
        switch (b.kind) {
            case "if_statement":
                return this.formatIfStatement(b as IfStatement);
            case "for_statement":
                return this.formatForStatement(b as ForStatement);
            case "switch_statement":
                return this.formatSwitchStatement(b as SwitchStatement);
            case "expression_statement":
                return {
                    kind: "expression_statement",
                    expression: this.formatExpression((b as ExpressionStatement).expression),
                    ...(b.asResult ? { asResult: b.asResult.resultName.name } : {}),
                };
            case "assignment_statement":
                return this.formatAssignmentStatement(b as AssignmentStatement);
            case "return_statement":
                return {
                    kind: "return_statement",
                    expression: this.formatExpression((b as ReturnStatement).expression),
                };
            case "return_error_statement":
                return {
                    kind: "return_error_statement",
                    value: this.formatExpression((b as ReturnErrorStatement).value),
                };
            case "check_block_statement":
                return {
                    kind: "check_block_statement",
                    resultName: (b as CheckBlockStatement).resultName.name,
                    ...((b as CheckBlockStatement).errorType
                        ? { errorType: (b as CheckBlockStatement).errorType!.name.name }
                        : {}),
                    body: this.formatBlockStatement((b as CheckBlockStatement).body),
                };
            case "forward_statement":
                return {
                    kind: "forward_statement",
                    resultName: (b as ForwardStatement).resultName.name,
                };
            case "variable_declaration_statement":
                return {
                    kind: "variable_declaration_statement",
                    name: b.name.name,
                    type: b.type.name.name,
                    ...(b.type.arrayLengths?.length ? { arrayLengths: b.type.arrayLengths } : {}),
                    ...(b.type.slice ? { slice: true } : {}),
                    value: this.formatExpression(b.value),
                    ...(b.asResult ? { asResult: b.asResult.resultName.name } : {}),
                };
        }
    }

    /** Formats a block as the list of its formatted statements. */
    formatBlockStatement(b: BlockStatement): any[] {
        return b.statements.map((x) => this.formatStatement(x));
    }

    /** Formats an identifier node. */
    formatIdentifier(i: Identifier) {
        return {
            kind: i.kind,
            name: i.name,
        };
    }

    /** Formats a type reference, including its name and resolved value. */
    formatType(t: Type): any {
        return {
            kind: t.kind,
            name: this.formatIdentifier(t.name),
            value: t.value,
            ...(t.arrayLengths?.length ? { arrayLengths: t.arrayLengths } : {}),
            ...(t.slice ? { slice: true } : {}),
            ...(t.variadic ? { variadic: true } : {}),
            ...(t.interfaceBounds?.length
                ? { interfaceBounds: t.interfaceBounds.map((x) => this.formatType(x)) }
                : {}),
            ...(t.typeParameters?.length
                ? { typeParameters: t.typeParameters.map((x) => this.formatType(x)) }
                : {}),
        };
    }

    /** Formats a function parameter as its name and type. */
    formatFunctionParameter(p: FunctionParameter) {
        return {
            name: this.formatIdentifier(p.name),
            type: this.formatType(p.type),
            ...(p.variadic ? { variadic: true } : {}),
        };
    }

    /** Formats a function declaration: signature, parameters, and body. */
    formatFunctionDeclaration(f: FunctionDeclaration) {
        const body = this.formatBlockStatement(f.body as BlockStatement);
        const parameters = f.parameters.map((x) => this.formatFunctionParameter(x));
        return {
            kind: "function_declaration",
            name: f.name,
            ...(f.typeParameters?.length
                ? { typeParameters: f.typeParameters.map((x) => this.formatType(x)) }
                : {}),
            ...(f.receiver ? { receiver: this.formatFunctionParameter(f.receiver) } : {}),
            parameters,
            returnTypes: f.returnTypes.map((x) => this.formatType(x)),
            errorType: f.errorTypes.map((x) => this.formatType(x)),
            ...(f.external ? { external: f.external } : {}),
            ...(f.exported ? { exported: true } : {}),
            body,
        };
    }

    /**
     * Formats a type declaration, dispatching on its `declKind` so each of the
     * three forms — `struct`, `enum`, `union` — renders its own body shape.
     */
    formatTypeDeclaration(d: TypeDeclaration): any {
        const base = {
            kind: "type_declaration",
            name: this.formatIdentifier(d.name),
            declKind: TypeDeclKind[d.declKind],
        };

        switch (d.declKind) {
            case TypeDeclKind.Struct: {
                const decl = d.declaration as StructDecl;
                return {
                    ...base,
                    fields: decl.fields.map((f) => ({
                        name: this.formatIdentifier(f.name),
                        type: this.formatType(f.type),
                    })),
                    ...(decl.implementedInterfaces?.length
                        ? {
                              implementedInterfaces: decl.implementedInterfaces.map((type) =>
                                  this.formatType(type),
                              ),
                          }
                        : {}),
                };
            }
            case TypeDeclKind.Enum: {
                const decl = d.declaration as EnumDecl;
                return {
                    ...base,
                    variants: decl.variants.map((v) => ({
                        name: v.name,
                        value: v.value,
                    })),
                };
            }
            case TypeDeclKind.Union: {
                const decl = d.declaration as UnionDecl;
                return {
                    ...base,
                    variants: decl.variants.map((v) => ({
                        type: this.formatType(v),
                    })),
                };
            }
        }
    }

    formatInterfaceMethod(method: InterfaceMethodRequirement) {
        return {
            kind: method.kind,
            name: this.formatIdentifier(method.name),
            ...(method.typeParameters?.length
                ? {
                      typeParameters: method.typeParameters.map((type) => this.formatType(type)),
                  }
                : {}),
            parameters: method.parameters.map((parameter) =>
                this.formatFunctionParameter(parameter),
            ),
            returnTypes: method.returnTypes.map((type) => this.formatType(type)),
            errorTypes: method.errorTypes.map((type) => this.formatType(type)),
        };
    }

    formatInterfaceDeclaration(declaration: InterfaceDeclaration) {
        return {
            kind: declaration.kind,
            name: this.formatIdentifier(declaration.name),
            methods: declaration.methods.map((method) => this.formatInterfaceMethod(method)),
            ...(declaration.exported ? { exported: true } : {}),
            ...(declaration.external ? { external: declaration.external } : {}),
        };
    }

    /** Formats a top-level declaration, dispatching on its `kind`. */
    formatDeclaration(d: Declaration) {
        switch (d.kind) {
            case "import_declaration": {
                const declaration = d as ImportDeclaration;
                return {
                    kind: declaration.kind,
                    ...(declaration.unsafe ? { unsafe: true } : {}),
                    ...(declaration.namespace
                        ? {
                              module: declaration.namespace.module.name,
                              ...(declaration.namespace.alias
                                  ? { alias: declaration.namespace.alias.name }
                                  : {}),
                          }
                        : { specifiers: declaration.specifiers.map((x) => x.name.name) }),
                    path: declaration.path,
                };
            }
            // case "variable_declaration_statement":
            //     return this.formatVar;
            case "function_declaration":
                return this.formatFunctionDeclaration(d as FunctionDeclaration);
            case "interface_declaration":
                return this.formatInterfaceDeclaration(d as InterfaceDeclaration);
            case "type_declaration":
                return this.formatTypeDeclaration(d as TypeDeclaration);
        }
    }

    /** Formats the whole module into a `{ file, declarations }` tree. */
    format() {
        const declarations = this.ast.declarations.map((x) => this.formatDeclaration(x));
        return {
            file: this.ast.fileName,
            declarations,
            ...(this.ast.exportModule ? { exportModule: this.ast.exportModule.name.name } : {}),
            ...(this.ast.ffiHeaders?.length ? { ffiHeaders: this.ast.ffiHeaders } : {}),
            ...(this.ast.ffiModuleName ? { ffiModule: this.ast.ffiModuleName } : {}),
            ...(this.ast.ffiLibraries?.length
                ? { ffiLibraries: this.ast.ffiLibraries.map(({ kind, path }) => ({ kind, path })) }
                : {}),
        };
    }

    /** Formats the module and prints it to stdout as YAML. */
    dump() {
        const formatResult = this.format();
        console.log(convert(formatResult));
    }
}
